lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -0,0 +1,1009 @@
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/SHA256.h"
+
+using namespace circt;
+using namespace firrtl;
+using hw::InnerRefAttr;
+
+/// This stores a mapping from a module name to every NLA that it particiapates
+/// in.
+using NLAMap = DenseMap<Attribute, std::vector<NonLocalAnchor>>;
+
+NLAMap createNLAMap(CircuitOp circuit) {
+  DenseMap<Attribute, std::vector<NonLocalAnchor>> nlaMap;
+  for (auto nla : circuit.getBody()->getOps<NonLocalAnchor>()) {
+    for (auto element : nla.namepath()) {
+      if (auto moduleRef = element.dyn_cast<FlatSymbolRefAttr>())
+        nlaMap[moduleRef.getAttr()].push_back(nla);
+      else
+        nlaMap[element.cast<InnerRefAttr>().getModule()].push_back(nla);
+    }
+  }
+  return nlaMap;
+}
+
+//===----------------------------------------------------------------------===//
+// Hashing
+//===----------------------------------------------------------------------===//
+
+llvm::raw_ostream &printHex(llvm::raw_ostream &stream,
+                            ArrayRef<uint8_t> bytes) {
+  // Print the hash on a single line.
+  return stream << format_bytes(bytes, llvm::None, 32) << "\n";
+}
+
+llvm::raw_ostream &printHash(llvm::raw_ostream &stream, llvm::SHA256 &data) {
+  auto string = data.result();
+  ArrayRef bytes(reinterpret_cast<const uint8_t *>(string.begin()),
+                 string.size());
+  return printHex(stream, bytes);
+}
+
+llvm::raw_ostream &printHash(llvm::raw_ostream &stream, std::string data) {
+  ArrayRef bytes(reinterpret_cast<const uint8_t *>(data.c_str()),
+                 data.length());
+  return printHex(stream, bytes);
+}
+
+struct StructuralHasher {
+  explicit StructuralHasher(MLIRContext *context) {
+    nonessentialAttributes.insert(StringAttr::get(context, "annotations"));
+    nonessentialAttributes.insert(StringAttr::get(context, "name"));
+    nonessentialAttributes.insert(StringAttr::get(context, "portAnnotations"));
+    nonessentialAttributes.insert(StringAttr::get(context, "portNames"));
+    nonessentialAttributes.insert(StringAttr::get(context, "portSyms"));
+    nonessentialAttributes.insert(StringAttr::get(context, "portTypes"));
+    nonessentialAttributes.insert(StringAttr::get(context, "sym_name"));
+    nonessentialAttributes.insert(StringAttr::get(context, "inner_sym"));
+  };
+
+  std::string hash(FModuleLike module) {
+    update(&(*module));
+    auto hash = sha.final().str();
+    reset();
+    return hash;
+  }
+
+private:
+  void reset() {
+    currentIndex = 0;
+    indexes.clear();
+    sha.init();
+  }
+
+  void update(const void *pointer) {
```

Comment:
This seems dangerous, what runtime-data is it capturing?

Reply:
Or rather, is there a narrow type that would be safer?

Reply:
This is for anytime I call `getOpaquePointer()`, which returns a `void *`.  RIght now this is capturing: `TypeID`, `Type`, `Attribute`, `OperationName`.  I could duplicate this function for each one to make it safer, would that be better?